dali/operators/audio/nonsilence_op.h

```diff
@@ -0,0 +1,202 @@
+ *
+ * If the length == 0, begin == 0
+ *
+ * Example:
+ * buffer: [0, 0, 0, 0, 50, 50, 0, 0]
+ * cutoff: 20
+ * return: {4, 2}
+ *
+ * @return (begin_idx, length)
+ */
+template<typename T>
+std::pair<int, int> LeadTrailThresh(span<const T> buffer, T cutoff) {
+  assert(buffer.size() > 0);
+  int begin = -1;
+  int end = buffer.size();
+  while (begin < end && buffer[++begin] < cutoff);  // NOLINT
+  if (begin == end) return {0, 0};  // Rest is silence
+  while (buffer[--end] < cutoff);  // NOLINT
+  return {begin, end - begin + 1};
+}
+
+
+template<typename InputType>
+void RunKernel(TensorView<StorageCPU, const InputType, 1> in, Tensor<CPUBackend> &out,
+               const kernels::signal::MovingMeanSquareArgs &mms_args) {
+  kernels::KernelContext kctx;
+  kernels::signal::MovingMeanSquareCpu<InputType> mms;
+  auto reqs = mms.Setup(kctx, in, mms_args);
+  out.Resize(reqs.output_shapes[0][0]);
+  auto tv = view_as_tensor<float>(out);
+  mms.Run(kctx, tv.template to_static<1>(), in, mms_args);
+}
+
+
+/**
+ * Performs nonsilent region detection for single sample
+ * @return (begin_idx, length)
+ */
+template<typename InputType>
+std::pair<int, int>
+DetectNonsilenceRegion(Tensor<CPUBackend> &intermediate_buffer, const Args<InputType> &args) {
+  RunKernel(args.input, intermediate_buffer, {args.window_length, args.reset_interval});
+  auto dbs = view_as_tensor<float>(intermediate_buffer);
+  kernels::signal::DecibelCalculator<float> dbc(10.f, args.reference_max ? max(dbs)
+                                                                         : args.reference_power);
+  return LeadTrailThresh(make_cspan(dbs.data, dbs.num_elements()), dbc.db2signal(args.cutoff_db));
+}
+
+}  // namespace detail
+
+template<typename Backend>
+class NonsilenceOperator : public Operator<Backend> {
+ public:
+  ~NonsilenceOperator() override = default;
+
+  DISABLE_COPY_MOVE_ASSIGN(NonsilenceOperator);
+
+ protected:
+  explicit NonsilenceOperator(const OpSpec &spec) :
+          Operator<Backend>(spec) {}
+
+
+  bool CanInferOutputs() const override {
+    return true;
+  }
+
+
+  void AcquireArgs(const OpSpec &spec, const workspace_t<Backend> &ws) {
+    this->GetPerSampleArgument(cutoff_db_, "cutoff_db", ws);
+    this->GetPerSampleArgument(reference_power_, "reference_power", ws);
+    window_length_ = spec.GetArgument<int>("window_length", &ws);
+    auto input_type = ws.template InputRef<Backend>(0).type().id();
+    // If input type is not floating point, there's no need for reset interval
+    reset_interval_ = IsFloatingPoint(input_type) ? spec.GetArgument<int>("reset_interval", &ws)
+                                                  : -1;
+  }
+
+
+  std::vector<float> cutoff_db_;
+  std::vector<float> reference_power_;
+  int window_length_ = -1;
+  int reset_interval_ = -1;
+
+  USE_OPERATOR_MEMBERS();
+};
+
+
+class NonsilenceOperatorCpu : public NonsilenceOperator<CPUBackend> {
```

Comment:
So can I instantiate `NonsilenceOperator<CPUBackend>`? What would be the difference of just having `NonsilenceOperator<CPUBackend>` ?

Reply:
You normally won't go with instantiating `NonsilenceOperator<CPUBackend>`. You'll do sth like:
```
class NonsilenceOperatorCpu : public NonsilenceOperator<CPUBackend>;
class NonsilenceOperatorGpu : public NonsilenceOperator<GPUBackend>;
DALI_REGISTER_OPERATOR(NonsilentRegion, NonsilenceOperatorCpu, CPU);
DALI_REGISTER_OPERATOR(NonsilentRegion, NonsilenceOperatorGpu, GPU);
```
The base class `NonsilenceOperator<T>` is to prevent boiler-plate duplication.

Reply:
Actually, instantiating Operator<CPUBackend> is what we normally do
We normally have something like:
```DALI_REGISTER_OPERATOR(NonsilentRegion, NonsilentRegion<CPUBackend>, CPU);```

Reply:
We do both ;)
```
DALI_REGISTER_OPERATOR(BrightnessContrast, BrightnessContrastCpu, CPU)
DALI_REGISTER_OPERATOR(Transpose, TransposeGPU, GPU);
```